druid/src/main/java/org/apache/calcite/adapter/druid/DruidConnectionImpl.java

```diff
@@ -338,8 +338,40 @@ private void parseField(List<String> fieldNames, List<ColumnMetaData.Rep> fieldT
       break;
     case VALUE_STRING:
     default:
-      rowBuilder.set(i, parser.getText());
-      break;
+      String s = parser.getText();
+      if (type != null) {
+        switch (type) {
+        case LONG:
+        case PRIMITIVE_LONG:
+        case SHORT:
+        case PRIMITIVE_SHORT:
+        case INTEGER:
+        case PRIMITIVE_INT:
+          if (s.equals("Infinity") || s.equals("-Infinity") || s.equals("NaN")) {
+            throw new RuntimeException("/ by zero");
+          }
+        case FLOAT:
+        case PRIMITIVE_FLOAT:
+        case PRIMITIVE_DOUBLE:
+        case NUMBER:
+        case DOUBLE:
+          if (s.equals("Infinity")) {
+            rowBuilder.set(i, Double.POSITIVE_INFINITY);
+            break;
```

Comment:
Thanks, I did not realize about the fall through. I guess it is matter of style, usually I prefer to nest switch statements within default case in this scenario, as I find it more readable, but this is valid too.
If you leave it as it is, maybe you could add ```//fallthrough``` at the end of the each such ```case``` to make it more evident, as you prefer.